rpcserver/router.go

```diff
@@ -687,6 +689,200 @@ func (server *routedBoltzServer) CreateReverseSwap(ctx context.Context, request
+		IsAuto:            isAuto,
+		ServiceFeePercent: utils.Percentage(chainPair.Fees.Percentage),
+	}
+
+	parseDetails := func(details *boltz.ChainSwapData, currency boltz.Currency) (*database.ChainSwapData, error) {
+		swapData := &database.ChainSwapData{
+			Id:                 response.Id,
+			Currency:           currency,
+			Amount:             details.Amount,
+			TimeoutBlockHeight: details.TimeoutBlockHeight,
+			Tree:               details.SwapTree.Deserialize(),
+			LockupAddress:      details.LockupAddress,
+		}
+		var walletName string
+		if currency == pair.From {
+			swapData.PrivateKey = refundPrivateKey
+			swapData.Address = request.GetRefundAddress()
+			walletName = request.GetFromWallet()
+		} else {
+			swapData.PrivateKey = claimPrivateKey
+			swapData.Address = request.GetClaimAddress()
+			walletName = request.GetToWallet()
+		}
+
+		if swapData.Address == "" {
+			wallet, err := server.onchain.GetAnyWallet(onchain.WalletChecker{
+				Name:     walletName,
+				Currency: pair.From,
+				EntityId: macaroons.EntityFromContext(ctx),
+			})
+			if err != nil {
+				return nil, fmt.Errorf("could not find wallet %s: %w", walletName, err)
+			}
+			id := wallet.GetWalletInfo().Id
+			swapData.WalletId = &id
+		} else {
+			if err := boltz.ValidateAddress(server.network, swapData.Address, currency); err != nil {
+				return nil, err
+			}
+		}
+
+		swapData.TheirPublicKey, err = btcec.ParsePubKey(details.ServerPublicKey)
+		if err != nil {
+			return nil, err
+		}
+
+		if currency == boltz.CurrencyLiquid {
+			swapData.BlindingKey, _ = btcec.PrivKeyFromBytes(details.BlindingKey)
+		}
+
+		if err := swapData.InitTree(currency == pair.To); err != nil {
+			return nil, err
+		}
+
+		if err := swapData.Tree.Check(boltz.ChainSwap, swapData.TimeoutBlockHeight, preimageHash); err != nil {
+			return nil, err
+		}
+
+		if err := swapData.Tree.CheckAddress(details.LockupAddress, server.network, swapData.BlindingPubKey()); err != nil {
+			return nil, err
+		}
+
+		return swapData, nil
+	}
+
+	chainSwap.ToData, err = parseDetails(response.ClaimDetails, pair.To)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	chainSwap.FromData, err = parseDetails(response.LockupDetails, pair.From)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	logger.Info("Verified redeem script and address of chainSwap " + chainSwap.Id)
+
+	err = server.database.CreateChainSwap(chainSwap)
+	if err != nil {
+		return nil, handleError(err)
+	}
+
+	/*
+		timeoutHours := boltz.BlocksToHours(response.TimeoutBlockHeight-blockHeight, pair.From)
+		ChainSwapResponse := &boltzrpc.CreateChainSwapResponse{
+			Id:                 chainSwap.Id,
+			Address:            response.Address,
+			ExpectedAmount:     int64(response.ExpectedAmount),
+			Bip21:              response.Bip21,
+			TimeoutBlockHeight: response.TimeoutBlockHeight,
+			TimeoutHours:       float32(timeoutHours),
+		}
+	*/
+
+	if !request.GetExternalPay() {
+		// TODO: custom block target?
```

Comment:
Maybe also custom fee? Or do we leave all those custom options to the user via external pay?